src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs

```diff
@@ -84,59 +84,77 @@ public ChunkedEncodingReadStream(HttpConnection connection) : base(connection)
                 return false;
             }
 
-            private async Task ConsumeChunkBytesAsync(ulong bytesConsumed, CancellationToken cancellationToken)
+            private Task ConsumeChunkBytesAsync(ulong bytesConsumed)
             {
                 Debug.Assert(bytesConsumed <= _chunkBytesRemaining);
                 _chunkBytesRemaining -= bytesConsumed;
-                if (_chunkBytesRemaining == 0)
+                return _chunkBytesRemaining != 0 ?
+                    Task.CompletedTask :
+                    ReadNextLineAndThrowIfNotEmptyAsync();
+            }
+
+            private async Task ReadNextLineAndThrowIfNotEmptyAsync()
+            {
+                _connection._allowedReadLineBytes = 2; // \r\n
+                if (!LineIsEmpty(await _connection.ReadNextLineAsync().ConfigureAwait(false)))
                 {
-                    _connection._allowedReadLineBytes = 2; // \r\n
-                    if (!LineIsEmpty(await _connection.ReadNextLineAsync(cancellationToken).ConfigureAwait(false)))
-                    {
-                        ThrowInvalidHttpResponse();
-                    }
+                    ThrowInvalidHttpResponse();
                 }
             }
 
             public override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
             {
                 ValidateBufferArgs(buffer, offset, count);
-                return ReadAsync(new Memory<byte>(buffer, offset, count)).AsTask();
+                return ReadAsync(new Memory<byte>(buffer, offset, count), cancellationToken).AsTask();
             }
 
-            public override async ValueTask<int> ReadAsync(Memory<byte> destination, CancellationToken cancellationToken = default)
+            public override async ValueTask<int> ReadAsync(Memory<byte> destination, CancellationToken cancellationToken)
             {
+                cancellationToken.ThrowIfCancellationRequested();
+
                 if (_connection == null || destination.Length == 0)
                 {
                     // Response body fully consumed or the caller didn't ask for any data
                     return 0;
                 }
 
-                if (_chunkBytesRemaining == 0)
+                CancellationTokenRegistration ctr = _connection.RegisterCancellation(cancellationToken);
```

Comment:
Can we defer this to after the _chunkBytesRemaining == 0 test?

Reply:
Will that buys us much?  If chunkBytesRemaining == 0 then we're going to do TryGetNextChunkAsync, for which we want the registration in place as it potentially does a read, and if chunkBytesRemaining != 0, then we're going to do a ReadAsync, for which we want the registration in place.  Am I misunderstanding?

Reply:
Yes, you're right -- I missed that we need the registration for the subsequent ReadAsync.  

That said, in the scenario where we have buffered bytes that we can use to satisfy the read, we will unnecessarily register for the cancellation token.  Not sure this matters in practice, and it's not obvious how to fix this in the code, given how HttpConnection.ReadAsync works with buffering and cancellation.  Still, might be worth a comment.

Reply:
If the registration cost proves troublesome, we could potentially add a "try" fast path, where we try to parse/read from what's in the buffer, and then fall back to doing all of this if we were unsuccessful.  I'm hopeful that with https://github.com/dotnet/coreclr/pull/12819 the overheads here shouldn't be too impactful.  And the costs are also largely nops if the cancellation token is None / the default; unfortunately that's never the case when going through HttpClient, but going through HttpMessageInvoker it can be.

Reply:
Ok, seems reasonable to me.

Reply:
One option I thought of is to use a custom awaiter, e.g.
```C#
private struct WithConfigureAwaitAndCancellation<T> : ICriticalNotifyCompletion
{
    private readonly ConfiguredValueTaskAwaitable<T>.ConfiguredValueTaskAwaiter _awaiter;
    private readonly CancellationToken _token;
    private readonly HttpConnection _connection;
    private CancellationTokenRegistration _ctr;

    public WithConfigureAwaitAndCancellation(
        ConfiguredValueTaskAwaitable<T>.ConfiguredValueTaskAwaiter awaiter,
        HttpConnection connection,
        CancellationToken cancellationToken)
    {
        _awaiter = awaiter;
        _connection = connection;
        _token = cancellationToken;
        _ctr = default;
    }

    public WithConfigureAwaitAndCancellation<T> GetAwaiter() => this;

    public void UnsafeOnCompleted(Action continuation)
    {
        _ctr = _connection.RegisterCancellation(_token);
        _awaiter.UnsafeOnCompleted(continuation);
    }

    public void OnCompleted(Action continuation) => UnsafeOnCompleted(continuation);

    public bool IsCompleted => _awaiter.IsCompleted;

    public T GetResult()
    {
        _ctr.Dispose();
        return _awaiter.GetResult();
    }
}
```
and use that when awaiting on each of the operations.  That way, we only end up registering with the token if we actually yield (if IsCompleted returns false).  I'm not going to do that now, but we can try it out subsequently to see if it makes a difference.